### PR TITLE
sewing-kit-koa now checks for matching locales

### PR DESCRIPTION
**Pull request by GoodForOneFare**
## Description
When provided with manifests containing a `locale` attribute, sewing-kit-koa will now attempt to match a user request to a set of assets with that locale's translations embedded into its JavaScript.

This is a companion to the react-i18n PR that supports per-locale builds: https://github.com/Shopify/quilt/pull/1197.  The advantages of this are:
* Bundlers like webpack don't get bloated with metadata about where to find async translations
* Users don't have to download multiple sets of translations

## Type of change

- [x] `@shopify/sewing-kit-koa` Minor: New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have added a changelog entry, prefixed by the type of change noted above

### CODE REVIEW

File: packages/sewing-kit-koa/src/manifests.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+export class Manifests {
+  manifestPath: string;
+  private resolvedManifestEntry?: Manifest;
+
+  constructor(manifestPath: string | null = null) {
+    this.manifestPath = manifestPath || DEFAULT_MANIFEST_PATH;
+  }
+
+  async getResolvedManifest(
```
Comment (TheMallen) [inline packages/sewing-kit-koa/src/manifests.ts]: nit: `Manifests#getResolvedManifest` seems a bit redundant, we could shorten to something like `Manifests#get` or `Manifests#resolve`

File: packages/sewing-kit-koa/src/manifests.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+const DEFAULT_MANIFEST_PATH = 'build/client/assets.json';
+const MULTI_ASYNC_LANGUAGE_MANIFESTS = 'noLocales';
+
+export interface ResolveOptions {
+  locale?: string | undefined;
+}
+
+export class Manifests {
+  manifestPath: string;
```
Comment (TheMallen) [inline packages/sewing-kit-koa/src/manifests.ts]: same with some of this stuff, since this new class is well named and specific I think we can make a lot of stuff less wordy in here by deleting the word `manifest`

File: packages/sewing-kit-koa/src/manifests.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      if (!manifests) {
+        manifests = [];
+        acc.set(locale, manifests);
+      }
+      manifests.push(manifest);
+    });
+  } else {
+    acc.get(MULTI_ASYNC_LANGUAGE_MANIFESTS)!.push(manifest);
+  }
+  // console.log(JSON.stringify([...acc], null, 2));
```
Comment (TheMallen) [inline packages/sewing-kit-koa/src/manifests.ts]: 🔥

File: packages/sewing-kit-koa/src/manifests.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    manifest.identifier = {target: manifest.name};
+  }
+  return manifest;
+}
+
+function groupManifestsByLocale(
+  acc: Map<string, ConsolidatedManifest>,
+  manifest: Manifest,
+) {
+  const locales = manifest.identifier!.locales;
```
Comment (TheMallen) [inline packages/sewing-kit-koa/src/manifests.ts]: is this really optional? If not why don't we change the type (it's from this library isn't it?) if so, we should maybe do an assertion above with a nice error message if it's not there?

File: packages/sewing-kit-koa/src/assets.ts [hunk 5]
```diff
@@ -52,24 +31,22 @@ enum AssetKind {
...
-const DEFAULT_MANIFEST_PATH = 'build/client/assets.json';
-
 export default class Assets {
   assetPrefix: string;
   userAgent?: string;
-  manifestPath: string;
+  private manifests: Manifests;
   private resolvedManifestEntry?: Manifest;
 
   constructor({assetPrefix, userAgent, manifestPath}: Options) {
```
Comment (TheMallen) [inline packages/sewing-kit-koa/src/assets.ts]: Given that we create[ one instance of this per request](https://github.com/Shopify/quilt/blob/master/packages/sewing-kit-koa/src/middleware.ts#L36) is there a reason we don't add `locale` to the class itself? Then we could remove the params all over the place for `locale` and just use the class attribute. Feels weird that we have a mix of params and class attributes for things that are per request.

**Comment by GoodForOneFare** [comment]
>  but on an app like web with many languages we might want to be aware of how it changes, especially once there are 100+. Do we have numbers or anything for that?

The cost is all up front: read a larger manifest on first request, and sort it into buckets.  For every other request:
- There's one extra map operation to see if the user's locale exists
- For locale / no locale flows, exactly the same number of browser checks are performed as before

So it feels to me like the runtime cost is one extra map access and maybe an extra function call ‾\\_(ツ)_/‾

File: packages/sewing-kit-koa/src/manifests.ts (hunk 1 above)
Comment (GoodForOneFare) [reply packages/sewing-kit-koa/src/manifests.ts]: 👍 Good call.  I went with resolve.

File: packages/sewing-kit-koa/src/manifests.ts (hunk 2 above)
Comment (GoodForOneFare) [reply packages/sewing-kit-koa/src/manifests.ts]: Chopped some stuff out 👍 Probably still a bit wordy in places, but I tried to let go 😄

File: packages/sewing-kit-koa/src/assets.ts (hunk 5 above)
Comment (GoodForOneFare) [reply packages/sewing-kit-koa/src/assets.ts]: Yah, this is nasty ☹️  It's an artefact of [how userAgent is available when the middleware creates Assets](https://github.com/Shopify/quilt/blob/5ba6cd69793a071950467c5f09c4dee9e93d15d0/packages/sewing-kit-koa/src/middleware.ts#L35-L45), but our apps tend to fetch locale info via GraphQL.

Locales are a niche feature right now, and I figure it's better to just add this verbose version of things before we put any more work into the API.

File: packages/sewing-kit-koa/src/manifests.ts (hunk 4 above)
Comment (GoodForOneFare) [reply packages/sewing-kit-koa/src/manifests.ts]: It's optional for now because `identifier`'s presence depends on an unreleased version of sewing-kit 😅 

The mandatory part can come 🔜 🤞
